packages/keyed/src/index.ts

```diff
@@ -0,0 +1,146 @@
+      }
+
+      const result = new Array<U>(list.length);
+
+      // fast path for new create or after fallback
+      const fb = prev.get(FALLBACK);
+      if (!prev.size || fb) {
+        fb?.dispose();
+        prev.delete(FALLBACK);
+        for (let i = 0; i < list.length; i++) {
+          const item = list[i];
+          const key = keyFn(item);
+          addNewItem(result, item, i, key);
+        }
+        return result;
+      }
+
+      const prevKeys = new Set(prev.keys());
+
+      for (let i = 0; i < list.length; i++) {
+        const item = list[i];
+        const key = keyFn(item);
+        prevKeys.delete(key);
+        const lookup = prev.get(key);
+
+        if (lookup) {
+          result[i] = lookup.mapped;
+          lookup.setIndex?.(i);
+          lookup.setItem(() => item);
+        } else addNewItem(result, item, i, key);
+      }
+
+      prevKeys.forEach(key => {
+        prev.get(key)?.dispose();
+        prev.delete(key);
+      });
+
+      return result;
+    })
+  );
+
+  function addNewItem(list: U[], item: T, i: number, key: K): void {
+    createRoot(dispose => {
+      const [getItem, setItem] = createSignal(item);
+      const save = { setItem, dispose } as Save;
+      if (mapFn.length > 1) {
+        const [index, setIndex] = createSignal(i);
+        save.setIndex = setIndex;
+        save.mapped = mapFn(getItem, index);
+      } else save.mapped = (mapFn as any)(getItem);
+      prev.set(key, save);
+      list[i] = save.mapped;
+    });
+  }
+}
+
+/**
+ * creates a list of elements from the input `each` list
+ *
+ * it receives a map function as its child that receives a **list item signal** and **index signal** and returns a JSX-Element; if the list is empty, an optional fallback is returned:
+ * ```tsx
+ * <Key each={items()} by={item => item.id} fallback={<div>No items</div>}>
+ *   {(item, index) => <div data-index={index()}>{item()}</div>}
+ * </Key>
+ * ```
+ *
+ * prop `by` can also be an object key:
+ * ```tsx
+ * <Key each={items()} by="id">
+ * ```
+ *
+ * @see https://github.com/davedbase/solid-primitives/tree/main/packages/keyed#Key
+ */
+export function Key<T, U extends JSX.Element>(props: {
+  each?: readonly T[] | null | false;
+  by: ((v: T) => any) | keyof T;
+  fallback?: U;
+  children: (v: Accessor<T>, i: Accessor<number>) => U;
+}): Accessor<U[]> {
+  const fallback = props.fallback ? () => props.fallback as U : undefined;
+  const { by } = props;
+  const key = typeof by === "function" ? by : (v: T) => v[by];
+  const mapped = mapKey<T, U, any>(() => props.each, key, props.children, { fallback });
+  return createMemo(mapped);
+}
+
+/**
+ * Causes the children to rerender when the `on` changes.
+ * @see https://github.com/davedbase/solid-primitives/tree/main/packages/refs#Rerun
+ */
+export const Rerun: Component<{ on: any }> = props => {
+  const key = createMemo(() => props.on);
+  return createMemo(() => {
+    key();
+    return props.children;
+  });
+};
```

Comment:
Hi!
Do we really need 2 memos to implement this feature?
Personally, I would rather go with something like this:
``` 
export const Rerun: Component<{ on: Parameters<typeof on>[0] }> = props => {
  return createMemo(on(props.on, () => props.children));
};
```
With this, we will follow typings of the first argument of the `on` function for overall consistency (and also it will be possible to pass an array of dependencies).
Otherwise, if you do not like this approach for the `on` prop and still want to pass it as a value I suggest the second approach to remove additional memo:
```
export const Rerun: Component<{ on: any }> = props => {
  const key = typeof props.on === "function" ? props.on : () => props.on;
  return createMemo(on(key, () => props.children));
};
```
With this approach, it will be possible to pass both values and functions as in your initial suggestion.

Reply:
Interesting, I knew there are some optimisations to be made, so thank you for pointing that one out. Or rather two.
I think I went with a separate memo for the key to not rerun the components, if the value didn't actually change, but has caused computation to run for some reason. Similarly how the Show component memoizes the condition.
But I wonder if that may be actually what the user wants, like you could have some special signal with `equals: false` used to re-render the component, where the actual value doesn't matter.
I'm down to changing it though, the future will tell what's the better approach anyway.
As for enabling functions to be passed to `on` I'm totally down. I've completely forgotten about that possibility. Thanks.

Reply:
Litte update:
```tsx
<Rerun on={[count, className]}>
  {([count, className]) => (
    <button class={className} onClick={() => setCount(p => ++p)}>
      {count}
    </button>
  )}
</Rerun>
```